spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/RevertDeleteProcedure.java

```diff
@@ -0,0 +1,196 @@
+      new StructType(
+          new StructField[] {
+            new StructField("added_files_count", DataTypes.LongType, false, Metadata.empty())
+          });
+
+  public static SparkProcedures.ProcedureBuilder builder() {
+    return new BaseProcedure.Builder<RevertDeleteProcedure>() {
+      @Override
+      protected RevertDeleteProcedure doBuild() {
+        return new RevertDeleteProcedure(tableCatalog());
+      }
+    };
+  }
+
+  private RevertDeleteProcedure(TableCatalog tableCatalog) {
+    super(tableCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    Identifier tableIdent = toIdentifier(args.getString(0), PARAMETERS[0].name());
+    long snapshotId = args.getLong(1);
+    boolean dryRun = !args.isNullAt(2) && args.getBoolean(2);
+
+    return modifyIcebergTable(
+        tableIdent,
+        table -> {
+          long count = revertDelete(table, snapshotId, dryRun);
+          InternalRow row = newInternalRow(count);
+          return new InternalRow[] {row};
+        });
+  }
+
+  public long revertDelete(Table table, long snapshotId, boolean dryRun) {
+    Snapshot snapshot = findSnapshot(table, snapshotId);
+
+    checkSnapshotIsDelete(table, snapshot);
+
+    checkForPreviousRevertDelete(table, snapshot.snapshotId());
+
+    AppendFiles append = table.newAppend();
+    long count = appendFilesPreviouslyRemoved(table, snapshot, append);
+
+    if (dryRun) {
+      LOG.debug("Dry run, would have appended {} files back to table {}", count, table.name());
+    } else {
+      append.commit();
+      LOG.debug("Committed {} files back to table {}", count, table.name());
+    }
+
+    return count;
+  }
+
+  private Snapshot findSnapshot(Table table, long snapshotId) {
+    Snapshot snapshot = table.snapshot(snapshotId);
+    if (snapshot == null) {
+      throw new IllegalArgumentException(
+          String.format("Snapshot %d in table %s was not found", snapshotId, table.name()));
+    }
+
+    if (!SnapshotUtil.isAncestorOf(table, snapshotId)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Snapshot %d in table %s is not an ancestor of the current table state",
+              snapshotId, table.name()));
+    }
+
+    return snapshot;
+  }
+
+  private void checkSnapshotIsDelete(Table table, Snapshot snapshot) {
+    // only allow DELETE operations...
+    if (!snapshot.operation().equals(DataOperations.DELETE)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Snapshot %d in table %s is not a delete operation",
+              snapshot.snapshotId(), table.name()));
+    }
+
+    // Ensure we don't have deltas...
+    String addedDeleteFiles =
+        snapshot.summary().getOrDefault(SnapshotSummary.ADDED_DELETE_FILES_PROP, "0");
```

Comment:
Do we also want to check that there were no deleted positional/equality delete files? Allowing a restore in that case would possibly change the data of the snapshot, so probably best to exclude that type of delete snapshot as well.